sale.sol

```diff
@@ -0,0 +1,153 @@
+  function transfer(address _to, uint256 _value) returns (bool success);
+
+  /// @notice send `_value` token to `_to` from `_from` on the condition it is approved by `_from`
+  /// @param _from The address of the sender
+  /// @param _to The address of the recipient
+  /// @param _value The amount of token to be transferred
+  /// @return Whether the transfer was successful or not
+  function transferFrom(address _from, address _to, uint256 _value) returns (bool success);
+
+  /// @notice `msg.sender` approves `_spender` to spend `_value` tokens
+  /// @param _spender The address of the account able to transfer the tokens
+  /// @param _value The amount of tokens to be approved for transfer
+  /// @return Whether the approval was successful or not
+  function approve(address _spender, uint256 _value) returns (bool success);
+
+  /// @param _owner The address of the account owning tokens
+  /// @param _spender The address of the account able to transfer the tokens
+  /// @return Amount of remaining tokens allowed to spent
+  function allowance(address _owner, address _spender) constant returns (uint256 remaining);
+
+  event Transfer(address indexed _from, address indexed _to, uint256 _value);
+  event Approval(address indexed _owner, address indexed _spender, uint256 _value);
+}
+
+contract Sale {
+  // We will pay funds to the association created.
+  address public beneficiary;
+
+  // Sale parameters
+  uint public fundingGoal;
+  uint public deadline;
+  uint public pricePerToken;
+  AbstractToken public rewardToken;
+
+  modifier afterDeadline() {
+    if (now < deadline) throw;
+    _;
+  }
+  
+  // Tracking variables
+  mapping(address => uint256) public contributionOf; // Track contributors
+  uint public amountRaised; //defaults to 0.
+  bool fundingGoalReached = false;
+  bool crowdsaleClosed = false;
+
+  // Events
+  event GoalReached(address beneficiary, uint amountRaised);
+  event FundsReceived(address backer, uint amount);
+  event FundsWithdrawal(address recipient, uint amount);
+  
+  function Crowdsale(
+    address beneficiary,
+    uint fundingGoalInEthers,
+    uint durationInMinutes,
+    uint etherCostPerToken,
+    AbstractToken rewardTokenAddress)
+  {
+    beneficiary = beneficiary;
+    fundingGoal = fundingGoalInEthers * 1 ether;
+    deadline = now + durationInMinutes * 1 minutes;
+    pricePerToken = etherCostPerToken * 1 ether;
+    rewardToken = AbstractToken(rewardTokenAddress);
+  }
+
+  // This function is called when the contract is paid ether.
+  function ()
+    payable
+  {
+    // Contribution possible?
+    // NOTE: what happens if contract is out of tokens?
+    if (crowdsaleClosed) throw;
+
+    uint amount = msg.value;
+    contributionOf[msg.sender] = amount;
+    amountRaised += amount;
+
+    // Send correct amount of tokens to the contributor.
+    // NOTE: Is this integer division?
+    rewardToken.transfer(msg.sender, amount / pricePerToken);
```

Comment:
Check the result that "transfer" returns.
From Solidity docs: "If you call a function via a message call but it does not finish properly (i.e. it runs out of gas, has no matching function, or throws an exception itself), except when a low level operation call, send, delegatecall or callcode is used. The low level operations never throw exceptions but indicate failures by returning false."